src/Thread.tsx

```diff
@@ -0,0 +1,55 @@
+import {defined} from './Base';
+import {fetchMessageIdsAndLabels, fetchMessagesByIds} from './Gapi';
+import {Message} from './Message';
+
+export class Thread {
+  private _rawThread: gapi.client.gmail.Thread;
+  private _messages: Message[] | undefined;
+  private _firstAndLastMessages: Message[] | undefined;
+
+  constructor(thread: gapi.client.gmail.Thread) {
+    this._rawThread = thread;
+    this._messages = undefined;
+    this._firstAndLastMessages = undefined;
+  }
+
+  get id(): string {
+    return defined(this._rawThread.id);
+  }
+
+  get messages(): Message[] {
+    return defined(this._messages);
+  }
+
+  get firstAndLastMessages(): Message[] | undefined {
```

Comment:
I shy away from using getters in general. In the calling location it's often surprising that it's calling a function. How about just making the fields public?

Reply:
I kinda prefer making these things immutable from outside the class.

Would just making these normal functions SGTM?